source/common/http/http2/codec_impl.cc

```diff
@@ -138,20 +138,14 @@ void ConnectionImpl::StreamImpl::encodeTrailersBase(const HeaderMap& trailers) {
 
 void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadata_map_vector) {
   ASSERT(parent_.allow_metadata_);
-
-  getMetadataEncoder().createPayload(metadata_map_vector);
-
-  // Estimates the number of frames to generate, and breaks the while loop when the size is reached
-  // in case submitting succeeds and packing fails, and we don't get error from packing.
-  const size_t frame_count = metadata_encoder_->frameCountUpperBound();
-  size_t count = 0;
-  // Keep submitting extension frames if there is payload left in the encoder.
-  while (metadata_encoder_->hasNextFrame() && count++ <= frame_count) {
-    submitMetadata();
-    parent_.sendPendingFrames();
+  MetadataEncoder& metadata_encoder = getMetadataEncoder();
+  if (!metadata_encoder.createPayload(metadata_map_vector)) {
+    return;
   }
-
-  ASSERT(!metadata_encoder_->hasNextFrame());
+  for (uint8_t flags : metadata_encoder.payloadFrameFlags()) {
+    submitMetadata(flags);
+  }
+  parent_.sendPendingFrames();
```

Comment:
is it important or a perf optimization that we submit N times and send once, rather than prior submit:send logic?
If so consider commenting

Reply:
Not important, anymore. This order just made more sense to me once I finished mentally untangling the submit operations from the send operations. (And it makes sure the sequence of those operations is always the same regardless of the network, where the bug happened because they were usually but not always interleaved - but the submits don't depend on the sends anymore so that doesn't matter.)